File: 프로그래머스/lv3/43238. 입국심사/입국심사.kt
class Solution {
    fun solution(n: Int, times: IntArray): Long {
        var answer = 0L
        var left = 0L
        var right = n.toLong() * times.maxOrNull()!!

        while (left <= right) {
            val mid = (left + right) / 2

            if (times.fold(0L) { total, num -> total + mid / num } >= n) {
                answer = mid
                right = mid - 1
            } else {
                left = mid + 1
            }
        }

        return answer
    }
}

File: 프로그래머스/lv3/43238. 입국심사/입국심사.py
def solution(n, times):
    answer = 0
    left, right = 0, max(times) * n
    
    while left <= right:
        mid = (left + right) // 2
        check = 0
        
        for time in times:
            check += mid // time
            
            if check >= n:
                break
    
        if check >= n:
            answer = mid
            right = mid - 1
        else:
            left = mid + 1
    
    return answer